File: 24.cc
#include <iostream>
#include <vector>
#include <sstream>
#include <algorithm>

using namespace std;

using pii = pair<int,int>;
using vpii = vector<pii>;

template <typename ... Params>
void print(const Params &... args)
{ ((cout << args << ' '), ...) << endl; }

int go1(const vpii & cmps, vpii & bridge) {

    int maxscore = 0;

    function<void(vpii &, int, int)> backtracking;
    backtracking = [&](vpii & seen, int port, int score) {
        if (score > maxscore) {
            maxscore = score;
            bridge = seen;
        }
        for (const auto & cmp: cmps) {
            int L = cmp.first, R = cmp.second;
            if (find(seen.begin(), seen.end(), cmp) == seen.end()\
            && (port == L || port == R)) {
                int otherport = (port == L) ? R : L;
                seen.push_back(cmp);
                backtracking(seen, otherport, score + L + R);
                seen.pop_back();
            }
        }
    };

    vpii seen;
    backtracking(seen,0,0);
    return maxscore;
}

int go2(const vpii & cmps, vpii & bridge) {

    int maxlen = 0;
    int maxscore = 0;

    function<void(vpii &, int, int, int)> backtracking;
    backtracking = [&](vpii & seen, int port, int len, int score) {
        if (len > maxlen || (len == maxlen && score > maxscore)) {
            maxlen = len;
            maxscore = score;
            bridge = seen;
        }

        for (const auto& cmp : cmps) {
            int L = cmp.first, R = cmp.second;
            if (find(seen.begin(), seen.end(), cmp) == seen.end() \
            && (port == L || port == R)) {
                int otherport = (port == L) ? R : L;
                seen.push_back(cmp);
                backtracking(seen, otherport, len + 1, score + L + R);
                seen.pop_back();
            }
        }
    };

    vpii seen;
    backtracking(seen, 0, 0, 0);
    return maxscore;
}

string getpath(const vpii & path) {
    string res;
    int i = -1;
    while (++i < path.size())
        res += to_string(path[i].first) + "/" + \
            to_string(path[i].second) + \
            (i < path.size() - 1 ? "-": "");
    return res;
}

int main() {

    vpii cmps;
    string line;

    while (getline(cin, line)) {
        stringstream ss(line);
        int a, b;
        char _;
        ss >> a >> _ >> b;
        cmps.push_back({a, b});
    }

    vpii bridge1, bridge2;

    int p1 = go1(cmps, bridge1);
    print("part 1:", p1);

    int p2 = go2(cmps, bridge2);    
    print("part 2:", p2);

    print("path 1:",getpath(bridge1));
    print("path 2:",getpath(bridge2));
}

File: 24.py
cmps = []
for line in open(0).read().splitlines():
    a,b = list(map(int,line.split('/')))
    cmps.append((int(a),int(b)))

def go1(cmps):
    res = 0
    bridge = []
    def backtracking(seen,port,score):
        nonlocal bridge, res
        if res < score:
            res = score
            bridge = seen[:]
        for cmp in cmps:
            l,r = cmp
            if cmp not in seen and port in cmp:
                otherport = l if l != port else r
                seen.append(cmp)
                backtracking(seen, otherport, score + sum(cmp))
                seen.pop()
    backtracking([],0,0)
    return res, bridge

def go2(cmps):
    maxsize = 0
    maxscore = 0
    bridge = []
    def backtracking(seen,port,size,score):
        nonlocal bridge, maxsize, maxscore
        if maxsize <= size:
            maxsize = size
            maxscore = max(maxscore,score)
            bridge = seen[:]
        for cmp in cmps:
            l,r = cmp
            if cmp not in seen and port in cmp:
                otherport = l if l != port else r
                seen.append(cmp)
                backtracking(seen, otherport, size + 1, score + sum(cmp))
                seen.pop()
    backtracking([],0,0,0)
    return maxscore, bridge

p1,path = go1(cmps)
print('part 1:',p1)

p2,path = go2(cmps)
print('part 2:',p2)

print('path 1:','-'.join(['/'.join(list(map(str,_)))for _ in path]))
print('path 2:','-'.join(['/'.join(list(map(str,_)))for _ in path]))

